appinventor/buildserver/src/com/google/appinventor/buildserver/FormPropertiesAnalyzer.java

```diff
@@ -160,87 +158,62 @@ private static void collectComponentTypes(JSONObject componentProperties,
    * @return A mapping of component type names to sets of blocks used
    */
   public static Map<String, Set<String>> getComponentBlocksFromBlocksFile(String source) {
-    Map<String, Set<String>> result = new HashMap<>();
+    final Map<String, Set<String>> result = new HashMap<>();
     if (source == null || source.isEmpty()) {
       return result;
     }
     try {
-      DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
-      DocumentBuilder builder = factory.newDocumentBuilder();
-      Document doc = builder.parse(new ByteArrayInputStream(source.getBytes(StandardCharsets.UTF_8)));
-      NodeList blocks = doc.getElementsByTagName("block");
-      for (int i = 0; i < blocks.getLength(); i++) {
-        Element block = (Element) blocks.item(i);
-        String type = block.getAttribute("type");
-
-        // Skip non-component and disabled blocks
-        if (!isComponentBlock(type) || "true".equals(block.getAttribute("disabled"))) continue;
-
-        Element mutation = getMutation(block);
-        if (mutation != null) {  // Should always be true, in theory...
-          String componentType = mutation.getAttribute("component_type");
-          String blockName = getBlockName(type, mutation);
-          if (!result.containsKey(componentType)) {
-            result.put(componentType, new HashSet<String>());
+      XMLReader reader = XMLReaderFactory.createXMLReader();
+      reader.setContentHandler(new DefaultHandler() {
+        private int skipBlocksCounter = 0;
+        private String blockType;
+
+        @Override
+        public void startElement(String uri, String localName, String qName, Attributes attributes)
+            throws SAXException {
+          if ("block".equals(qName)) {
+            if ("true".equals(attributes.getValue("disabled"))) {
+              skipBlocksCounter = 1;
```

Comment:
I don't think that this is the right thing to do here. Rather, you should increment `skipBlocksCounter`. This is because it is possible to disable a block and then disable an ancestor, and when walking the tree you will reset the counter back to 1 when reaching the disabled descendant. I was able to construct a set of blocks using this technique that would include the SEND_SMS permission even though all the blocks look disabled in the editor.

<img width="342" alt="Screen Shot 2019-03-22 at 12 51 22 AM" src="https://user-images.githubusercontent.com/1331253/54801318-aadefd00-4c3c-11e9-85f2-a2e8f5d49ede.png">

Both the = block and the if block are disabled, but a compiled app will still have the permission as if the SendMessageDirect block is enabled.

Reply:
Thanks